### PR TITLE
ActionListopsWidget

### PR DESCRIPTION
**Pull request by matabele**
Introduction
---------------

You suggestions to modify the action-mangletags widget to function with lists in fields other than the 'tags' field was fairly easily implemented. However, the new functionality introduced complications, as it appears that many of the uses to which lists are put, involve maintaining the order of the list. 

At first I tried the direct approach, modifying the action-mangletags widget with a growing list of attributes to manipulate the target list. Ultimately my efforts appeared fruitless -- it would be near impossible to anticipate all of the uses to which lists might be put, and include options to allow for all of these uses.

I therefore, began experimenting with a technique I had used for my legacy MakeList widget, which saved the output of a filter expression to a field. In this case, I modified the widget to use the contents of the current list as input for the filter expression, then saved the modified list back to the field containing the list.

Initial results were encouraging; many list operations were possible using the existing set of filters. The main omissions were filters designed to append, remove and replace items in the list. I then set about writing filters to remedy these deficiencies.

Further Development
----------------------------

With the addition of a few filters, it became apparent that the technique provided an elegant alternative for the manipulation of lists. Few changes will be necessary for the basic action-listops widget, as the widget itself is simple -- the complexity is confined to the construction of associated filter expressions, and the more alternative filters there are available, the more easily are these filter expressions constructed.

After some testing of the widgets, I now have a working action-widget (action-listops widget), with an associated library of extended filters (x-listops filters), which is capable of implementing numerous list operations. I have written some documentation with respect the use of the widget -- this is available at: http://listops.tiddlyspot.com/

I think the widget is now ready to be exposed for thorough testing, where input from others can assist in polishing the code and documentation.

### CODE REVIEW

**Comment by tobibeer** [comment]
:+1:  great stuff @matabele ...I have the feeling this will be very useful for a lot of things

**Comment by tobibeer** [comment]
The devil's advocate role I'm playing every now and then... I'm wondering though, how about a mere **filter** parameter for the [ActionSetFieldWidget](http://tiddlywiki.com/#ActionSetFieldWidget) that would do the same thing, evaluate a filter and set the field to the output?

**Comment by matabele** [comment]
There's already a facility in the core for doing this -- have a look at the $filter attributes of the 'set' widget. I don't know what the intended usage was, but it may be used for this purpose. 

I don't like the idea of placing the full functionality of the action-listops widget into the action-setfield widget -- this would only serve to complicate the UX. Adding some of the functionality, perhaps just the $filter attribute might be passable -- much like I have done with my x-setrefs widget (also available on the same wiki.) 

I prefer separating functions by widget -- the listops widget is aimed at managing lists, the setfield widget is aimed at managing fields. Rather include the functionality of the delete-field widget into the setfield widget. This way, if you wish to manage fields -- you go with the setfield widget; iIf you wish to manage lists (including the 'tags' list) -- you go with the listops widget.

I have a preference for the SetTextReference function, and dedicate a special widget for setting the values of TextReferences; the aim of the x-setrefs widget is, therefore, the management of values of TextReferences. As an addon, I include a basic ability to append and remove tags -- this is a simple convenience, but should not interfere with the core functionality of the widget (in this case, no attributes are needed to add this facility.)

**Comment by Jermolene** [comment]
Hi @matabele thank you for this, and apologies for the late response.

The new approach looks pretty good. Can you add the necessary documentation? I think we'd be in a position to merge then.

**Comment by matabele** [comment]
OK -- will do.

**Comment by matabele** [comment]
Have now added documentation. I am unsure of some of the conventions involved for TW5.com (especially for the documentation of filter operators) -- may need some tweaking.

**Comment by matabele** [comment]
I would appreciate some feedback with respect applying this widget to real usage cases -- I can then amend the code to cover missing usage cases.

**Comment by inmysocks** [comment]
For a real use case: I am playing with adding snap.js to tiddlywiki and I want to be able to add polygons with an arbitrary number of vertices, the work flow goes:
1) Select that you want to add a polygon
2)Click 'add point'
3)Click on the SVG canvas to set the first point, this point gets added to a list
4)Repeat 2) and 3) until you have added all the points you want

After than I want to be able to edit the order of the points and add or remove a point to any position in the list. I should also be able to edit a point, so it would need to be able to replace a list item with another item as well as insert/remove them.

I think that the widget can do everything needed for this and I this use covers all the other use cases I have. I haven't tried to actually use the widget for this yet. I will probably do that sometime soon.

**Comment by inmysocks** [comment]
Oh, I guess the other thing is allowing non-unique items, but looking at things now I think that would be easy to do and should probably be worried about after the main function is taken care of. I think that either another input like $unique=false or a filter operator like unique:false[] that would take away the part of the core that imposes uniqueness would be the way to go. It should be a very simple change that doesn't affect anything else so this is just here as a reminder.

**Comment by tobibeer** [comment]
@inmysocks, I don't think there is **any** core filter that allows duplicates and, unless you desperately need them (meaning: you could not model your solution without duplicates), I would not recommend introducing them.

**Comment by inmysocks** [comment]
I have updated the svg example to use the list ops widget if you want a working example to show. Adding a polygon uses the list opts widget to add and remove points. Thank you for making this, it is very very useful.
http://ooktech.com/jed/ExampleWikis/SnapSVG/bob.html

@tobibeer Similar to how I had to change $tw.utils.parseStringArray() when I tried to make something for list operations using the code for tags, for filters the function that needs to be modified is $tw.utils.pushTop(). The change would be almost identical to the change to parseStringArray and it would maintain the same backwards compatibility. The reasoning for wanting to allow duplicates is the same as it was before. If you want a concrete example than for the things I am doing with snap.js, to describe an animation for an SVG image you could want it to translate, rotate then translate again. These steps are stored in a list because that is the best way to work with sets of variable length in tiddlywiki, you could easily end up with a sequence of steps like this: t100,100 r0,0,30 t100,100. The order is important and there are repeated steps. For editing purposes you want to be able to do things like insert and remove or replace items in the list. Any tool like what I had made before would be repeating functionality in a far less elegant way.

**Comment by tobibeer** [comment]
@inmysocks, thanks for the reminder. Forgot about those details. For instructions like routing it is imperative to have lists not sets, e.g. `go: N N S E N W`. So, I take it the listops filters will be the only ones that allow that? If that is so, they should perhaps all have the same prefix so that we can see that these (can?!?) all act on lists, not sets.

**Comment by inmysocks** [comment]
@tobibeer I don't think that the operators themselves need to make the distinction. As far as I can tell the filter operators don't explicitly remove duplicates. What happens is that they try to add each item that the operator returns, even duplicates, to the list of results using the pushTop function. That function will only add to the list if the item isn't already on the list. This way all of the filter operators maintain uniqueness by default without anything in the filter operator code itself.

What we would need to do is pass the uniqueness parameter to the filtertiddlers function and make that pass it to the compilefilter function which in turn gives it to the pushtop function. We should be able to do that and maintain perfect backwards compatibility and we can get everything figured out for the action-listops widget before worrying about it.

**Comment by tobibeer** [comment]
> As far as I can tell the filter operators don't explicitly remove duplicates.

Try **any** filters, starting with:

```
{{{ foo foo foo }}}
```

> pushTop ... This way all of the filter operators maintain uniqueness by default without anything in the filter operator code itself.

I'm quite certain that not every filter operator uses **pushTop**. Possibly, the filter syntax may be enhanced with a new directive to have **any** filter behave as if working with a list, not a set. This will require revising all filters.

For that purpose I have created a new ticket: #2097.

> pass it to the compileFilter function which in turn gives it to the pushTop function

Perhaps, that's the right spot.

However, how would `-[any-filter[foo]]` work? Remove all?

**Comment by inmysocks** [comment]
Yes, every filter removes duplicates the way things are made now, but that is part of the compileFilter function, not any of the operators themselves. From what I have seen no filter operator uses pushTop, but anything that handles filters or lists uses either pushTop or parseStringArray before returning the list and those do remove duplicates. That is why a fix would be simple because it can all be done there.

**Comment by matabele** [comment]
The $subfilter attribute modifies the filter by appending one or more runs -- and each invocation of a run enforces uniqueness. The append[] and prepend[] filter operators can add items without enforcing uniqueness, however, upon the next run, uniqueness is again re-imposed. 

This can not be fixed without modifying the core (removing any code that enforces uniqueness.) All that would be required to restore this behaviour is a "unique[]" filter expression to remove duplicates. This could then be appended to any filter expression requiring uniqueness.

**Comment by tobibeer** [comment]
> All that would be required to restore this behaviour is a "unique[]" filter expression to remove duplicates. This could then be appended to any filter expression requiring uniqueness.

I believe uniqueness must always be presumed given rather than called for. Using a filter operator for this will not work as it is the very mechanism that joins filter runs that defines how the operation unfolds.

Please take a look at #2097. Here I suggest to define a prefix for the entire filterString that would trigger **list**- vs **set**-mode, set-mode being (the sensible) default.

If it was needed to actually be able to specify list-mode on the filter-run level, the syntax could be amended like so:

```
{{{ [tag[foo]] +[operator:sufix[operand]+] }}}
```

Notice the `+` inbetween the closing square brackets. (Added this to #2097).

**Comment by inmysocks** [comment]
Yes, that is why I said to give some sort of parameter. In the same way as #2027 the behaviour changes when given an extra parameter, if you leave it out than nothing changes from how it works now. Passing it like a filter operator could work, but if I understand the code correctly it would be simpler to pass it as a separate parameter (like <$list filter=somefilter unique=false/>) so that we don't have to touch the filter parsing code.

**Comment by matabele** [comment]
As you mentioned in #2097 -- assuming uniqueness introduces difficulties when removing items. If the uniqueness only gets imposed last, this isn't so much of a problem -- finish processing the list, then remove duplicates. 

This could be done using the second syntax option [filter[expression]+] vs. [filter[expression]] -- if the trailing + sign is missing --> remove duplicates, else leave the list as is.

**Comment by tobibeer** [comment]
> This could be done using the second syntax option [filter[expression]+] vs. [filter[expression]] -- if the trailing + sign is missing --> remove duplicates, else leave the list as is.

Yes, that was the idea at the level of joining a filter run with its input titles.

**Comment by inmysocks** [comment]
I just ran into a situation where being able to swap two items on a list would be useful. So take a list 1 2 3 4 5 and swap 2 and 5 to get 1 5 3 4 2. I think you could use something like replace for it. I am not sure what should happen if one of the items that are supposed to be swapped doesn't exist. No change may be the best way to go.

**Comment by inmysocks** [comment]
The widget doesn't handle spaces in list items.

**Comment by tobibeer** [comment]
@inmysocks, can you be more specific (example), please?

**Comment by inmysocks** [comment]
Here is an example: http://ooktech.com/jed/ExampleWikis/SnapSVG/bob.html#Make%20Group%20-%20Non-tags%20field%20example

It occurs to me that this may not be completely unexpected behavior because the operators are made to work on lists. Even if that is the case I think that we should do something to help add list items with spaces because '+[append<SomeVariable>]' always treats SomeVariable as a list.

**Comment by tobibeer** [comment]
@inmysocks, I think you're using it wrong.

Firstly, you don't need to specify $tiddler if you act on the current tiddler.

Most importantly, you need to wrap your title in square brackets, e.g.

```
\define bracketed() [[$(CurrentEiddler)$]]

<$action-listops $list='elements' $subfilter='+[append<bracketed>]'/>
```

**Comment by matabele** [comment]
The append[] and prepend[] operators may be used with a reference to a list -- in this way, titles with spaces can be included in the items to be added: 

<$action-listops $list="elements" $subfilter ="+[append{!!mylist}]"/>

Items may be selected from this list to append to the current list by adding a suffix and prefix:

$subfilter ="+[append:3{!!mylist}]" -- selects the first 3 items from 'mylist'

$subfilter ="+[!append:3{!!mylist}]" -- selects the last 3 items from 'mylist'

**Comment by inmysocks** [comment]
I really think that there should be some way to directly add titles with spaces. Could we add something like a title suffix like $subfilter="+[append:title[something with spaces]]"? This widget is really useful but I keep running into this problem and finding ways to add [[ and ]] can be awkward at times.

**Comment by matabele** [comment]
I have used standard filter syntax -- the inability to include a title with spaces directly in the operand is common to all filter operators. Another separator would be necessary, as double brackets [[ ]] aren't correctly parsed in filter operands.

A 'swap' might be done with a pair of move[] operators. I'm not sure how to specify the second item for a swap[] operator -- perhaps an index would suffice -- like the move[] operator -- as in +[swap:3[marker]]

**Comment by Jermolene** [comment]
Hi @matabele is the issue here that it is not possible to directly pass a list as the operand of a filter operator unless the members of the list do not contain spaces?

If so, we could consider extending the syntax to fix it. In particular, we could have a new style of inner brackets that indicates a list that quotes with double square brackets. For example:

```
+[append(HelloThere [[Title with spaces]])]
```

The implementation would be pretty simple: a matter of allowing round brackets as a synonym for the inner square brackets of a filter operand.

File: editions/tw5.com/tiddlers/filters/allafter.tid [hunk 1]
```diff
@@ -0,0 +1,17 @@
+caption: allafter
+created: 20151017145021839
+creator: matabele
```
Comment (Jermolene) [inline editions/tw5.com/tiddlers/filters/allafter.tid]: Hi @matabele we avoid the creator/modifier fields in tw5.com; the idea is that there's a single, unified editorial voice.

**Comment by matabele** [comment]
Hi Jeremy

The syntax +[append[one [[and two]] three]] clearly can't be parsed correctly -- your suggestion of +[append(one [[and two]] three)] seems to provide a neat solution.

**Comment by tobibeer** [comment]
> `[append(one [[and two]] three)]`

:+1: I very much like this. It appears safe to assume that we're never going to use closing round brackets in literal titles, thus breaking the expression. For that case we can resort to using a variable instead.

**Comment by Jermolene** [comment]
>  It appears safe to assume that we're never going to use closing round brackets in literal titles, thus breaking the expression. For that case we can resort to using a variable instead

We could perhaps also additionally support double round brackets:

```
[append((one a(1) [[and two]] three))]
```

File: editions/tw5.com/tiddlers/filters/allafter.tid (hunk 1 above)
Comment (matabele) [inline editions/tw5.com/tiddlers/filters/allafter.tid]: Many thanks -- noted. Should I now submit a new pull request with creator and modifier fields removed?

Comment (Jermolene) [inline editions/tw5.com/tiddlers/filters/allafter.tid]: Hi @matabele no worries, I removed them myself: c2105de49c0eacaead3244b628f3c64301d3420e

Comment (matabele) [inline editions/tw5.com/tiddlers/filters/allafter.tid]: OK -- seen. I am also inquiring with respect the general case. To modify code, I now submit a new pull request from a new branch?

Comment (Jermolene) [inline editions/tw5.com/tiddlers/filters/allafter.tid]: Hi @matabele yes, that's right - and do feel free to send further updates if you see any opportunities for improvement.

**Comment by matabele** [comment]
@inmysocks 

There's another way to add titles, without the use of append[] or prepend[] -- simply list the titles to add to the current list before the filter run. This method allows titles with spaces, or may reference a field or macro containing a title with spaces -- like so:

`$subfilter="title [[another title]] {{!!mytitle}} +[filter[expression]]"`

**Pull request merged**